lib-js/compiler.ts

```diff
@@ -0,0 +1,177 @@
+        command,
+        [...initArgs, ...args],
+        {
+          cwd: path,
+          encoding: "utf-8",
+          stdio: ["ignore", "pipe", "pipe"],
+        }
+      )
+      if (error) {
+        throw new Error(`Failed to run executable compiler: ${error}`)
+      }
+      return { stdout, stderr, status }
+    },
+  }
+}
+
+export class DartCompiler implements Compiler {
+  private readonly stdin: Writable
+  private readonly stdout: AsyncGenerator<string>
+  private readonly stderr: AsyncGenerator<string>
+
+  private constructor(
+    dart: ChildProcessWithoutNullStreams,
+    private readonly initArgs: string[] = []
+  ) {
+    this.stdin = dart.stdin
+    this.stdout = DartCompiler.toChunks(dart.stdout)
+    this.stderr = DartCompiler.toChunks(dart.stderr)
+    this.initArgs = initArgs
+  }
+
+  /**
+   * Create a dart-sass compiler from the repo given by the path.
+   */
+  static async fromRepo(path: string, initArgs: string[] = []) {
+    return new DartCompiler(await this.createProcess(path), initArgs)
+  }
+
+  async compile(path: string, opts: string[]) {
+    this.stdin.write(`!cd ${path}\n`)
+    this.stdin.write([...this.initArgs, ...opts].join(" ") + "\n")
+    return {
+      stdout: (await this.stdout.next()).value,
+      stderr: (await this.stderr.next()).value,
+      status: +(await this.stdout.next()).value,
+    }
+  }
+  /**
+   * Create a child process that uses the dart-sass repo at the path,
+   * and compiles the files piped to stdin.
+   */
+  private static async createProcess(repoPath: string) {
+    if (!fs.existsSync(path.resolve(repoPath, "bin/sass.dart"))) {
+      throw new Error(`${repoPath} is not a valid Dart Sass repository`)
+    }
+    const dartFile = `
+// @dart=2.9
+import "dart:convert";
+import "dart:io";
+
+import "${repoPath}/bin/sass.dart" as sass;
+
+main() async {
+  await for (var line in new LineSplitter().bind(utf8.decoder.bind(stdin))) {
+    if (line.startsWith("!cd ")) {
+      Directory.current = line.substring("!cd ".length);
+      continue;
+    }
+
+    try {
+      await sass.main(line.split(" ").where((arg) => arg.isNotEmpty).toList());
+    } catch (error, stackTrace) {
+      stderr.writeln("Unhandled exception:");
+      stderr.writeln(error);
+      stderr.writeln(stackTrace);
+      exitCode = 255;
+    }
+
+    stdout.add([0xFF]);
+    stdout.write(exitCode);
+    stdout.add([0xFF]);
+    stderr.add([0xFF]);
+    exitCode = 0;
+  }
+}`
+    const dartFilename = path.resolve(os.tmpdir(), "sass-wrapper.dart")
+    await fs.promises.writeFile(dartFilename, dartFile, { encoding: "utf-8" })
+    const child = child_process.spawn("dart", [
+      "--enable-asserts",
+      `--packages=${repoPath}/.packages`,
+      dartFilename,
+    ])
+    // TODO figure out how to delete the file without crashing the process
+    // fs.unlinkSync(dartFilename)
+    // child.on("spawn", () => {
+    //   fs.unlinkSync(dartFilename)
+    // })
```

Comment:
@nex3 your other commend got marked outdated, but I tried tempy and it runs into the same problem—node exits with an `EPIPE` if the file is deleted while the spawned process is still existent.

I can think of a couple ways we can resolve this:

1. still use tempy to generate a file name but just let it live in the tmp directory
2. make an explicit `cleanup` method for the compiler to call at the end of the test run.
3. Do more research and figure out how to delete the file.

Reply:
I wonder why this worked in Ruby but not in Node...

Could you add an `exit` listener that only unlinks the file once the process is done?

Reply:
That works! Thanks for the idea